src/ring_theory/dedekind_domain.lean

```diff
@@ -0,0 +1,76 @@
+/-
+Copyright (c) 2020 Kenji Nakagawa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenji Nakagawa, Anne Baanen, Filippo A. E. Nuccio
+-/
+import ring_theory.fractional_ideal
+import ring_theory.ideal.over
+import ring_theory.discrete_valuation_ring
+import order.zorn
+
+namespace ring
+
+/-- A ring `R` has Krull dimension at most one if all nonzero prime ideals are maximal. -/
+def dimension_le_one (R : Type*) [comm_ring R] : Prop :=
+∀ p ≠ (⊥ : ideal R), p.is_prime → p.is_maximal
+
+end ring
+
+open ideal ring
+
+lemma is_principal_ideal_ring.dimension_le_one (R : Type*) [integral_domain R]
+  [is_principal_ideal_ring R] :
+  ring.dimension_le_one R :=
+λ p nonzero prime, by { haveI := prime, exact is_prime.to_maximal_ideal nonzero }
+
+variables {R K : Type*}
+
+lemma integral_closure.dimension_le_one [comm_ring R] [nontrivial R] [integral_domain K] [algebra R K]
+  (h : dimension_le_one R) :
+  dimension_le_one (integral_closure R K) :=
+begin
+  intros p ne_bot prime,
+  haveI := prime,
+  refine integral_closure.is_maximal_of_is_maximal_comap p (h _ (integral_closure.comap_ne_bot ne_bot) _),
+  apply is_prime.comap
+end
+
+/-- A Dedekind domain is a nonfield that is Noetherian, integrally closed, and has Krull dimension exactly one (`not_is_field` and `dimension_le_one`).
+This is the default implementation, but there are equivalent definitions,
+`is_dedekind_domain_dvr` and `is_dedekind_domain_inv`.
+TODO: Prove that these are actually equivalent definitions.
+-/
+class is_dedekind_domain (R : Type*) [integral_domain R] : Prop :=
+(not_is_field : ¬ is_field R)
+(is_noetherian_ring : is_noetherian_ring R)
+(dimension_le_one : dimension_le_one R)
+(is_integrally_closed : integral_closure R (fraction_ring R) = ⊥)
```

Comment:
I had to change the class of `R` to be an `integral_domain`, because `fraction_ring R` uses this assumption. I expect after #3942 gets merged, we can make it a `comm_ring` again.

Reply:
I've kicked #3942 on the queue, so let's wait for that one. After that, I think we can merge this, ok?

Reply:
Sounds good. I'm still (slowly :( other obligations keep interfering this week) doing my own bit of reviewing though.

Reply:
#3942 was merged, was this able to be changed back?

Reply:
If it is possible to do so, I don't think we should change it to `comm_ring` because I don't think any of the conditions are strong enough to imply it is an `integral_domain` (although I'm not entirely sure for the integrally closed condition). Either way, I would guess that the proof of integrally closed for some concrete ring is going to be much nicer if you already have it. I might be entirely wrong, but this is just my two cents.

Reply:
Checking the literature, it looks like most, if not all definitions assume `R` is an integral domain, so let's follow that and keep `integral_domain R`. That is presumably the reason it's called a Dedekind *domain*, not a Dedekind *ring*. :)

(I think the trivial ring is a counterexample showing the Dedekind conditions are not enough to imply it's a domain.)